subprojects/model-groovy/src/integTest/groovy/org/gradle/cli/TaskOptionsSpec.groovy

```diff
@@ -0,0 +1,104 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.cli
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class TaskOptionsSpec extends AbstractIntegrationSpec {
+    def "can use a built-in option"() {
+        when:
+        buildScript """
+            abstract class MyTask extends DefaultTask {
+                @Option(option="profile", description="foobar")
+                void setProfile(String value) {
+                    System.out.println("profile=" + value)
+                }
+            }
+
+            tasks.register('mytask', MyTask.class)
+        """
+
+        then:
+        succeeds "mytask", "--profile"
+    }
+
+    def "can use a task option"() {
+        when:
+        buildScript """
+            abstract class MyTask extends DefaultTask {
+                @Option(option="profile", description="foobar")
+                @Optional
+                @Input
+                abstract Property<String> getProfile()
+
+                @TaskAction
+                void run() {
+                    if (getProfile().isPresent()) {
+                        logger.lifecycle("profile=" + getProfile().get())
+                    }
+                }
+            }
+
+            tasks.register('mytask', MyTask.class)
+        """
+
+        then:
+        succeeds "--", "mytask", "--profile", "myvalue"
+        output.contains "profile=myvalue"
+    }
+
+    def "task options apply to most recent task"() {
+        when:
+        buildScript """
+            abstract class MyTaskA extends DefaultTask {
+                @Option(option="profile", description="foobar")
+                @Optional
+                @Input
+                abstract Property<String> getProfile()
+
+                @TaskAction
+                void run() {
+                    if (getProfile().isPresent()) {
+                        logger.lifecycle("profile=" + getProfile().get())
+                    }
+                }
+            }
+
+            abstract class MyTaskB extends DefaultTask {
+                @Option(option="profile", description="foobar")
+                @Optional
+                @Input
+                abstract Property<String> getProfile()
+
+                @TaskAction
+                void run() {
+                    if (getProfile().isPresent()) {
+                        logger.lifecycle("profile=" + getProfile().get())
+                    }
+                }
+            }
+
+            tasks.register('mytaskA', MyTaskA.class)
+            tasks.register('mytaskB', MyTaskB.class)
+        """
+
+        then:
+        succeeds "--", "mytaskA", "--profile", "myvalueA", "mytaskB", "--profile", "myvalueB"
+        output.contains "Aprofile=myvalueA"
+        output.contains "Bprofile=myvalueB"
+    }
+}
```

Comment:
Finally, we'll need to add a couple tests for tasks defined in `buildSrc` and in included builds, ensure sure the `--` works the same way for them (probably worth mentioning briefly whether it does or doesn't in the docs).  I suspect it will, but am far from certain.

The trick to this is the `file` method, which uses wherever the test project is created as the root and allows you to create files relative to it.  So you could create a task in the `/buildSrc` dir like:

```groovy
file('buildSrc/src/main/java/MyTask') << """
public class MyTask extends DefaultTask {
  // task body ...
}
"""
```

and use similar calls to setup another project dir, add a buildscript to it, and make it an `includeBuild` via `settings.gradle` in the top-level build.

It's pretty rare to only need to modify a single project file in an integration test, actually, so knowing how to do this is definitely worthwhile.

Reply:
Added, thanks for the explanation. I'm just using basic "does the task option work" cases here -- let me know if you have something more robust in mind.